File: go/workflows/order_workflow_scenarios.go
package workflows

import (
	"fmt"
	"temporal-order-management/activities"
	"temporal-order-management/app"
	"temporal-order-management/messages"
	"time"

	"github.com/google/uuid"
	"go.temporal.io/api/enums/v1"
	"go.temporal.io/sdk/converter"
	"go.temporal.io/sdk/temporal"
	"go.temporal.io/sdk/workflow"
)

const (
	BUG        = "OrderWorkflowRecoverableFailure"
	CHILD      = "OrderWorkflowChildWorkflow"
	NEXUS      = "OrderWorkflowNexusOperation"
	SIGNAL     = "OrderWorkflowHumanInLoopSignal"
	UPDATE     = "OrderWorkflowHumanInLoopUpdate"
	VISIBILITY = "OrderWorkflowAdvancedVisibility"
)

var orderStatusKey = temporal.NewSearchAttributeKeyKeyword("OrderStatus")

func OrderWorkflowScenarios(ctx workflow.Context, args converter.EncodedValues) (output *app.OrderOutput, err error) {
	var input app.OrderInput
	err = args.Get(&input)
	if err != nil {
		return nil, fmt.Errorf("failed to decode arguments: %w", err)
	}

	name := workflow.GetInfo(ctx).WorkflowType.Name
	logger := workflow.GetLogger(ctx)
	logger.Info("Dynamic Order workflow started", "type", name, "orderId", input.OrderId)

	activityOptions := workflow.ActivityOptions{
		StartToCloseTimeout: 5 * time.Second,
		RetryPolicy: &temporal.RetryPolicy{
			InitialInterval:    1 * time.Second,
			BackoffCoefficient: 2.0,
			MaximumInterval:    30 * time.Second,
		},
	}
	ctx = workflow.WithActivityOptions(ctx, activityOptions)

	localActivityOptions := workflow.LocalActivityOptions{
		StartToCloseTimeout: 5 * time.Second,
	}
	laCtx := workflow.WithLocalActivityOptions(ctx, localActivityOptions)

	// Create saga to manage order compensations
	var saga app.Saga
	defer func() {
		if err != nil {
			disconnectedCtx, _ := workflow.NewDisconnectedContext(ctx)
			saga.Compensate(disconnectedCtx)
		}
	}()

	// Expose progress as query
	progress, err := messages.SetQueryHandlerForProgress(ctx)
	if err != nil {
		return nil, err
	}

	// Get items
	items := app.Items{}
	err = workflow.ExecuteLocalActivity(laCtx, activities.GetItems).Get(ctx, &items)
	if err != nil {
		return nil, err
	}

	updateProgress("Check Fraud", progress, 0, ctx, 0)

	// Check fraud
	err = workflow.ExecuteActivity(ctx, activities.CheckFraud, input).Get(ctx, nil)
	if err != nil {
		return nil, err
	}

	updateProgress("Prepare Shipment", progress, 25, ctx, 1)

	// Prepare shipment
	saga.AddCompensation(activities.UndoPrepareShipment, input)
	err = workflow.ExecuteActivity(ctx, activities.PrepareShipment, input).Get(ctx, nil)
	if err != nil {
		return nil, err
	}

	updateProgress("Charge Customer", progress, 50, ctx, 1)

	// Charge customer
	saga.AddCompensation(activities.UndoChargeCustomer, input)
	err = workflow.ExecuteActivity(ctx, activities.ChargeCustomer, input, name).Get(ctx, nil)
	if err != nil {
		return nil, err
	}

	updateProgress("Ship Order", progress, 75, ctx, 3)

	if BUG == name {
		// Simulate bug
		panic("Simulated bug - fix me!")
	}

	if SIGNAL == name {
		// Await signal message to update address
		logger.Info("Waiting up to 60 seconds for updated address")
		var updateInput messages.UpdateOrderInput
		c := messages.GetSignalChannelForUpdateOrder(ctx)
		ok, _ := c.ReceiveWithTimeout(ctx, time.Minute, &updateInput)
		if ok {
			input.Address = updateInput.Address
		}
	}

	if UPDATE == name {
		// Await update message to update address
		logger.Info("Waiting up to 60 seconds for updated address")
		updatedAddress, err := messages.SetUpdateHandlerForUpdateOrder(ctx)
		if err != nil {
			return nil, err
		}
		ok, _ := workflow.AwaitWithTimeout(ctx, time.Minute, func() bool {
			return *updatedAddress != ""
		})
		if ok {
			input.Address = *updatedAddress
		}
	}

	// Ship order items
	var shipFutures []workflow.Future
	for _, item := range items {
		logger.Info("Shipping item " + item.Description)
		shipFutures = append(shipFutures, shipItemAsync(ctx, input, item, name))
	}

	// Wait for all items to ship
	for _, f := range shipFutures {
		err = f.Get(ctx, nil)
		if err != nil {
			return nil, err
		}
	}

	updateProgress("Order Completed", progress, 100, ctx, 0)

	// Generate trackingId
	trackingId := uuid.New().String()
	output = &app.OrderOutput{
		TrackingId: trackingId,
		Address:    input.Address,
	}

	return output, nil
}

func updateProgress(orderStatus string, progress *int, value int, ctx workflow.Context, seconds int) {
	sleep(ctx, seconds, progress, value)
	if VISIBILITY == workflow.GetInfo(ctx).WorkflowType.Name {
		workflow.UpsertTypedSearchAttributes(ctx, orderStatusKey.ValueSet(orderStatus))
	}
}

func shipItemAsync(ctx workflow.Context, input app.OrderInput, item app.Item, name string) workflow.Future {
	logger := workflow.GetLogger(ctx)
	var f workflow.Future

	shippingInput := app.ShippingInput{
		Order: input,
		Item:  item,
	}

	if CHILD == name {
		// execute an async child wf to ship the item
		cwo := workflow.ChildWorkflowOptions{
			WorkflowID:        fmt.Sprintf("shipment-%v-%v", input.OrderId, item.Id),
			ParentClosePolicy: enums.PARENT_CLOSE_POLICY_TERMINATE,
		}
		ctx = workflow.WithChildOptions(ctx, cwo)
		f = workflow.ExecuteChildWorkflow(ctx, ShippingWorkflow, shippingInput)
		logger.Info("Started Child Workflow: " + cwo.WorkflowID)
	} else if NEXUS == name {
		client := workflow.NewNexusClient(app.GetEnv("TEMPORAL_NEXUS_SHIPPING_ENDPOINT", "shipping-endpoint"), app.ShippingServiceName)

		fut := client.ExecuteOperation(ctx, app.ShippingOperationName, shippingInput, workflow.NexusOperationOptions{})
		f = fut

		var exec workflow.NexusOperationExecution
		fut.GetNexusOperationExecution().Get(ctx, &exec)
		logger.Info("Started Nexus Operation: " + exec.OperationToken)
	} else {
		// execute an async activity to ship the item
		f = workflow.ExecuteActivity(ctx, activities.ShipOrder, shippingInput)
		logger.Info("Started Activity: ShipOrder ")
	}
	return f
}

File: ruby/workflows/order_workflow_scenarios.rb
require_relative 'order_workflow'
require 'temporalio/error'
require_relative 'shipping_child_workflow'

module Workflows
  class OrderWorkflowScenarios < OrderWorkflow
    workflow_dynamic

    BUG = "OrderWorkflowRecoverableFailure"
    CHILD = "OrderWorkflowChildWorkflow"
    SIGNAL = "OrderWorkflowHumanInLoopSignal"
    UPDATE = "OrderWorkflowHumanInLoopUpdate"
    VISIBILITY = "OrderWorkflowAdvancedVisibility"

    attr_accessor :progress, :updated_address, :retry_policy

    def initialize
      super
      @updated_address = nil
    end

    def execute(input)
      input = Models::OrderInput.new(input['OrderId'], input['Address']) if input.is_a?(Hash)
      workflow_type = Temporalio::Workflow.info.workflow_type
      logger.info("Dynamic Order workflow started, type = #{workflow_type}, orderId = #{input.order_id}")

      compensations = []

      order_items = Temporalio::Workflow.execute_activity(
        Activities::GetItemsActivity,
        start_to_close_timeout: 5
      )

      update_progress("Check Fraud", 0, 0)

      Temporalio::Workflow.execute_activity(
        Activities::CheckFraudActivity,
        input,
        start_to_close_timeout: 5,
        retry_policy: @retry_policy
      )

      update_progress("Prepare Shipment", 25, 1)

      compensations << Activities::UndoPrepareShipmentActivity
      Temporalio::Workflow.execute_activity(
        Activities::PrepareShipmentActivity,
        input,
        start_to_close_timeout: 5,
        retry_policy: @retry_policy
      )

      update_progress("Charge Customer", 50, 1)

      begin
        compensations << Activities::UndoChargeCustomerActivity
        Temporalio::Workflow.execute_activity(
          Activities::ChargeCustomerActivity,
          input, workflow_type,
          start_to_close_timeout: 5,
          retry_policy: @retry_policy
        )
      rescue StandardError => ex
        logger.error("Failed to charge customer: #{ex.message}")
        compensations.reverse.each do |comp|
          logger.error("Compensating: #{comp}")
          Temporalio::Workflow.execute_activity(
            comp,
            input,
            start_to_close_timeout: 10,
            retry_policy: @retry_policy
          )
        end
        raise ex
      end

      update_progress("Ship Order", 75, 3)

      raise "Simulated bug - fix me!" if workflow_type == BUG

      wait_for_updated_address_or_timeout(input) if [SIGNAL, UPDATE].include?(workflow_type)

      #TODO These is not working asynchronously
      activity_handles = order_items.map do |item|
        logger.info("Shipping item: #{item.description}")
        if workflow_type == CHILD
          Temporalio::Workflow::Future.new do
            Temporalio::Workflow.execute_child_workflow(
              Workflows::ShippingChildWorkflow,
              input, item,
              id: "shipment-#{input.order_id}-#{item.id}",
              parent_close_policy: :terminate
            )
          end
        else
          Temporalio::Workflow::Future.new do
            Temporalio::Workflow.execute_activity(
              Activities::ShipOrderActivity,
              input, item,
              start_to_close_timeout: 5,
              retry_policy: @retry_policy
            )
          end
        end
      end

      # Wait for all futures to complete
      results = activity_handles.map(&:wait)

      update_progress("Order Completed", 100, 0)

      tracking_id = Temporalio::Workflow.random.uuid
      Models::OrderOutput.new(tracking_id, input.address).deep_camelize_keys
    end

    def wait_for_updated_address_or_timeout(input)
      logger.info("Waiting up to 60 seconds for updated address")
      begin
        Temporalio::Workflow.timeout(60) { Temporalio::Workflow.wait_condition { @updated_address } }
        input.address = @updated_address
      rescue Timeout::Error
        logger.info("Updated address was not received within 60 seconds")
      end
    end

    def update_progress(order_status, progress, sleep_time)
      @progress = progress
      sleep(sleep_time) if sleep_time > 0
      if workflow_type == VISIBILITY
        logger.error("Updating search attributes with order status: #{order_status}")
        key = Temporalio::SearchAttributes::Key.new(
          "OrderStatus",
          Temporalio::SearchAttributes::IndexedValueType::KEYWORD
        )

        update = key.value_set(order_status)

        Temporalio::Workflow.upsert_search_attributes(update)
      end
    end

    workflow_query(name: 'getProgress')
    def query_progress
      @progress
    end

   workflow_signal(name: 'UpdateOrder')
    def update_order(update_input)
      logger.info("Received update order signal with address: #{update_input}")
      @updated_address = update_input['Address']
    end

    # TODO This won't work. Only one of these can be decorated with a specific name, unlike other SDKs where "UpdateOrder" is the name for all three
    # Only one of these will work at a time, and do so by making its name 'UpdateOrder' as that's what the UI calls. Not going to change all the otehr SDKs just for Ruby, which is still in Alpha
    workflow_update(name: 'UpdateOrderUpdate')
    def update_order_update(update_input)
      logger.info("Received update order update with address: #{update_input['Address']}")
      @updated_address = update_input['Address']
      "Updated address: #{@updated_address}"
    end

    workflow_update_validator(:update_order_update)
    def update_order_update_validator(update_input)
      puts("Validating address: #{update_input}")
      unless update_input['Address'][0] =~ /\d/
        logger.info("Rejecting order update, invalid address: #{update_input['Address']}")
        raise Temporalio::Error::ApplicationError.new("Address must start with a digit", type: "invalid-address")
      end
      logger.info("Order update address is valid: #{update_input['Address']}")
    end

    private

    def workflow_type
      Temporalio::Workflow.info.workflow_type
    end
  end
end
